File: src/main/java/com/baulsupp/oksocial/stackexchange/StackExchangeAuthInterceptor.java
package com.baulsupp.oksocial.stackexchange;

import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.credentials.CredentialsStore;
import com.baulsupp.oksocial.credentials.OSXCredentialsStore;
import com.baulsupp.oksocial.secrets.Secrets;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class StackExchangeAuthInterceptor implements AuthInterceptor<StackExchangeCredentials> {
  private final CredentialsStore<StackExchangeCredentials> credentialsStore =
      CredentialsStore.create(new StackExchangeServiceDefinition());
  private StackExchangeCredentials credentials = null;

  public static final String NAME = "stackexchange";

  @Override public String name() {
    return NAME;
  }

  @Override
  public Response intercept(Chain chain) throws IOException {
    Request request = chain.request();

    if (credentials() != null) {
      String token = credentials().serverToken;

      request =
          request.newBuilder().addHeader("Authorization", "Token " + token).build();
    }

    return chain.proceed(request);
  }

  public StackExchangeCredentials credentials() {
    if (credentials == null) {
      credentials = credentialsStore.readDefaultCredentials();
    }

    return credentials;
  }

  @Override
  public CredentialsStore<StackExchangeCredentials> credentialsStore() {
    return credentialsStore;
  }

  public boolean supportsUrl(HttpUrl url) {
    String host = url.host();

    return StackExchangeUtil.API_HOSTS.contains(host);
  }

  @Override
  public void authorize(OkHttpClient client) {
    System.err.println("Authorising StackExchange API");

    String clientId = Secrets.prompt("StackExchange Client Id", "stackexchange.clientId", false);
    String clientSecret = Secrets.prompt("StackExchange Client Secret", "stackexchange.clientSecret", true);
    Set<String> scopes = new HashSet<>(
        Arrays.asList(Secrets.prompt("Scopes", "stackexchange.scopes", false).split(",")));

    StackExchangeCredentials newCredentials =
        LoginAuthFlow.login(client, clientId, clientSecret, scopes);
    CredentialsStore<StackExchangeCredentials> stackOverflowCredentialsStore =
        new OSXCredentialsStore<>(new StackExchangeServiceDefinition());
    stackOverflowCredentialsStore.storeCredentials(newCredentials);
  }
}

File: src/main/kotlin/com/baulsupp/okurl/services/stackexchange/StackExchangeAuthInterceptor.kt
package com.baulsupp.okurl.services.stackexchange

import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.okurl.authenticator.AuthInterceptor
import com.baulsupp.okurl.authenticator.ValidatedCredentials
import com.baulsupp.okurl.credentials.ServiceDefinition
import com.baulsupp.okurl.credentials.TokenValue
import com.baulsupp.okurl.kotlin.query
import com.baulsupp.okurl.secrets.Secrets
import com.baulsupp.okurl.services.stackexchange.model.MeResponse
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response

class StackExchangeAuthInterceptor : AuthInterceptor<StackExchangeToken>() {
  override val serviceDefinition = object : ServiceDefinition<StackExchangeToken> {
    override fun apiHost() = "api.stackexchange.com"

    override fun serviceName() = "StackExchange API"

    override fun shortName() = "stackexchange"

    override fun parseCredentialsString(s: String): StackExchangeToken {
      val parts = s.split(":".toRegex(), 2).toTypedArray()
      return StackExchangeToken(parts[0], parts[1])
    }

    override fun formatCredentialsString(credentials: StackExchangeToken) =
      credentials.accessToken + ":" + credentials.key

    override fun apiDocs() = "https://api.stackexchange.com/docs"

    override fun accountsLink() = "http://stackapps.com/apps/oauth"
  }

  override suspend fun intercept(chain: Interceptor.Chain, credentials: StackExchangeToken): Response {
    var request = chain.request()

    val newUrl = request.url()
      .newBuilder()
      .addQueryParameter("access_token", credentials.accessToken)
      .addQueryParameter("key", credentials.key)
      .build()

    request = request.newBuilder().url(newUrl).build()

    return chain.proceed(request)
  }

  override suspend fun validate(client: OkHttpClient, credentials: StackExchangeToken): ValidatedCredentials {
    return ValidatedCredentials(
      client.query<MeResponse>(
        "https://api.stackexchange.com/2.2/me?site=stackoverflow",
        TokenValue(credentials)
      ).items.firstOrNull()?.display_name
    )
  }

  override suspend fun authorize(
    client: OkHttpClient,
    outputHandler: OutputHandler<Response>,
    authArguments: List<String>
  ): StackExchangeToken {

    val clientId = Secrets.prompt("StackExchange Client Id", "stackexchange.clientId", "", false)
    val clientSecret = Secrets.prompt(
      "StackExchange Client Secret", "stackexchange.clientSecret",
      "", true
    )
    val clientKey = Secrets.prompt("StackExchange Key", "stackexchange.key", "", false)
    val scopes = Secrets.promptArray(
      "Scopes", "stackexchange.scopes", listOf(
        "read_inbox",
        "no_expiry",
        "write_access",
        "private_info"
      )
    )

    return StackExchangeAuthFlow.login(client, outputHandler, clientId, clientSecret, clientKey, scopes)
  }
}
